extras/rya.manual/src/site/markdown/rya-streams.md

```diff
@@ -0,0 +1,385 @@
+
+### Historic Statement Processing ###
+As currently implemented, a program needs to write new Statements to the core
+Rya data store as well as the input topic for Rya Streams. Every time a query
+is added to Rya Streams for processing, it will only see newly added Statements.
+We could write a Kafka Connect Rya source that import historic data into the 
+statements topic so that it, too, will be processed.
+
+### PCJ Maintenance ### 
+The Rya Streams system is capable of creating PCJ results over streamed 
+statements, but there isn't anything written to import those results from the
+QueryResult topic into the source Rya's PCJ index. We could use Kakfa Connect
+to write those binidng sets into the index.
+
+### Query Manager Durability ###
+The Query Manager daemon is a single point of failure for this system. If it
+is not running, then queries can not be processed. This problem could be
+solved in a few different ways depending on the size of the deployed system.
+If multiple machines are available, but only a small amount of data/number of
+queries need to be processed, then failing over to a secondary Query Manager
+would work fine. If a cluster solution is being used, then the Query Manager
+could be refactored to run in something like YARN. It then becomes the
+cluster's responsibility to start up processing on a new machine when the
+application dies.
+
+### Query Change Log Reverse Compatibility ###
+We use Java Serialization to write ChangeLogEntry<QueryChange> objects to the 
+KafkaQueryChangeLog. This will cause problems if the schema of that object is 
+updated between versions. A serialization framework (Avro, Protobuf, etc) that 
+supports reverse compatible models needs to be used instead to move this feature 
+out of Beta.
+
+### Exactly Once Kafka Streams Processing ###
+The version of Kafka that we are using for this system does not guarantee
+exactly once execution within Kafka Streams. It only guarantees at least once
+execution. This means a query may process the same statement more than once.
+If this happens, then repeat results may be written to the query's results
+topic. Kafka 1.0.0+, however, does support exactly once execution. Upgrading
+Kafka would allow us to fix this problem.
+
+### Query Processing Scalability ###
+The Query Manager executes all query topologies with a single thread per query 
+on the same JVM. We suspect this will only work well when there are a limited 
+number of queries that need to be processed and all of those queries don't need
+to process a large volume of statements.
+
+Kafka Streams usually scales out by copying the processing topology to many
+machines who do the same processing, but over different hunks of work that need
+to be processed. Each streams worker handles a subset of the topic's partitions. 
+This works great so long as all data that is required to process the worker's
+portion of the processing  is located within a single partition of the topic. 
```

Comment:
Remove extra space between "processing is"

Reply:
Done.